File: agents/luna/prompts/prompt_generator.py
# ===================================================================
# 🌙 Dynamic Prompt Generator for Luna Analysis AI
# ===================================================================
#
# 작성자: Luna (Jr. TRANG)
# 대상: KODA (개발)
# 목적: 필요에 따라 프롬프트를 동적으로 생성
#
# 사용 위치: Mission Control의 #luna-analysis 채널
# 호출 방식: prompt = generator.generate("배추", context_data)
#
# ===================================================================

import json
from datetime import datetime
from typing import Dict, Optional

PROMPT_TEMPLATES = {
    "배추": """당신은 Mulberry Lab의 배추 전문 분석가, Luna(배추)입니다.

[정체성]
- 역할: 배추 공구의 성과 분석 및 전략 수립 전문가
- 대상: {팀}
- 톤: 전문적, 데이터 중심, 실무 지향

[책임]
1. 배추 공구의 참여율 분석
   - 전주 대비 증감 분석
   - 신규 참여자 추이
   - 재참여율 평가

2. 가격 경쟁력 평가
   - KOSTAT 시장 도매가 대비
   - 소비자 절감액 계산
   - 농민 수익성 평가

3. 품질 & 만족도
   - 농민 만족도 세부 분석
   - 배송 완료율 평가
   - 취소율 / 반품율 분석

4. 다음주 전략 제안
   - 공급량 조정 방안
   - 가격 조정 필요성
   - 마케팅 집중 영역

[분석 대상 데이터]
{데이터}

[분석 목표]
{목표}

분석을 시작하세요.""",

    "위험": """당신은 Mulberry Lab의 위험 관리 전문가, Luna(위험)입니다.

[정체성]
- 역할: 비정상 신호 감지 및 즉시 보고
- 대상: {팀}

[실시간 모니터링 데이터]
{데이터}

[현재 상황]
{상황}

이상 신호를 감지하세요.""",

    "팀": """당신은 Mulberry Lab의 팀 협력 AI, Luna(팀)입니다.

[정체성]
- 역할: Sr. TRANG Manager와 팀의 의사결정 지원
- 대상: {팀}

[분석 데이터]
{데이터}

[팀의 질문]
{질문}

팀과 함께 의사결정하세요.""",

    "admin": """당신은 Mulberry Lab의 CEO 참모, Luna(Admin)입니다.

[정체성]
- 역할: CEO re.eul의 신속한 의사결정 지원

[전략 데이터]
{데이터}

[CEO의 질문]
{질문}

신속하게 분석하세요.""",

    "농민": """당신은 Mulberry Lab의 농민 관계 전문가, Luna(농민)입니다.

[정체성]
- 역할: 농민 만족도, 신뢰도, 지원 방안 분석
- 대상: {팀}

[농민 데이터]
{데이터}

[분석 초점]
{목표}

농민과 함께 성장하세요.""",
}


class PromptGenerator:
    """
    동적 프롬프트 생성기

    사용 예시:
    >>> generator = PromptGenerator()
    >>> prompt = generator.generate(
    ...     template="배추",
    ...     context={
    ...         "팀": "Sr. TRANG Manager",
    ...         "목표": "다음주 전략 수립",
    ...         "데이터": "참여 118명...",
    ...         "기간": "2026-07-10 ~ 07-17"
    ...     }
    ... )
    """

    def __init__(self):
        self.templates = PROMPT_TEMPLATES
        self.generated_count = 0

    def generate(self, template: str, context: Dict) -> str:
        """
        동적으로 프롬프트를 생성합니다.

        Args:
            template (str): 템플릿 이름 ("배추", "위험", "팀", "admin", "농민")
            context (Dict): 플레이스홀더 값

        Returns:
            str: 최종 프롬프트 (Inkling에 전송 가능)
        """
        # Step 1: 템플릿 검증
        if template not in self.templates:
            available = list(self.templates.keys())
            raise ValueError(f"❌ 템플릿 '{template}' 없음
사용 가능: {available}")

        # Step 2: 기본 템플릿 가져오기
        base_prompt = self.templates[template]

        # Step 3: 컨텍스트로 플레이스홀더 대체
        try:
            final_prompt = base_prompt.format(**context)
        except KeyError as e:
            raise ValueError(f"❌ 필수 키 누락: {e}")

        # Step 4: 타임스탬프 추가
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        header = f"[프롬프트 생성: {template}]
[생성 시간: {timestamp}]
---
"

        # Step 5: 통계 업데이트
        self.generated_count += 1

        # Step 6: 최종 프롬프트 반환
        return header + final_prompt

    def get_template_info(self, template: str) -> Dict:
        """
        템플릿 정보 조회 (디버깅용)
        """
        info_map = {
            "배추": {"이름": "배추 전문가", "용도": "배추 공구 성과 분석 & 전략", "대상": "Sr. TRANG + 팀"},
            "위험": {"이름": "위험 관리 전문가", "용도": "비정상 신호 감지 & 즉시 보고", "대상": "Sr. TRANG + CEO"},
            "팀": {"이름": "팀 협력 AI", "용도": "의사결정 지원 & 토론 활성화", "대상": "Sr. TRANG + Koda + Kbin + Malu"},
            "admin": {"이름": "CEO 참모", "용도": "신속한 의사결정 지원", "대상": "CEO re.eul"},
            "농민": {"이름": "농민 관계 전문가", "용도": "농민 만족도 & 신뢰도 분석", "대상": "Malu + Sr. TRANG"},
        }
        return info_map.get(template, {})

    def validate_context(self, template: str, context: Dict) -> bool:
        """
        컨텍스트 데이터 검증
        """
        required_keys = {"데이터", "목표"}
        additional = {
            "배추": {"팀"},
            "위험": {"상황"},
            "팀": {"질문"},
            "admin": {"질문"},
            "농민": {"팀"},
        }
        missing = required_keys - set(context.keys())
        if missing:
            raise ValueError(f"필수 키 누락: {missing}")
        if template in additional:
            extra = additional[template]
            missing_extra = extra - set(context.keys())
            if missing_extra:
                raise ValueError(f"[{template}] 필수 키 누락: {missing_extra}")
        return True

    def get_stats(self) -> Dict:
        """
        생성 통계 반환
        """
        return {
            "총_생성_횟수": self.generated_count,
            "사용_가능_템플릿": list(self.templates.keys()),
            "템플릿_개수": len(self.templates)
        }


# ===================================================================
# PART 3: 사용 예시 (테스트/개발용)
# ===================================================================

if __name__ == "__main__":
    generator = PromptGenerator()

    context_배추 = {
        "팀": "Sr. TRANG Manager",
        "목표": "다음주 배추 공구 전략 수립",
        "데이터": "참여자: 118명, 단가: 2850원/kg, 만족도: 87%",
        "기간": "2026-07-10 ~ 2026-07-17"
    }

    prompt_배추 = generator.generate("배추", context_배추)
    print(prompt_배추)
    print(generator.get_stats())


# ===================================================================
# PART 4: Mission Control 통합 코드
# ===================================================================
# routes/analysis.py에서:
#
# from prompt_generator import PromptGenerator
# generator = PromptGenerator()
#
# @app.post('/api/analysis')
# def analyze(request_data):
#     template = request_data.get("template")
#     context = request_data.get("context")
#     prompt = generator.generate(template, context)
#     response = inkling.analyze(prompt)
#     ChatChannel.post('#luna-analysis', response)
#     return {"success": True, "result": response}
#
# ===================================================================
# END OF FILE
# ===================================================================

File: mulberry-mission-control/utils/prompt-generator.js
/**
 * utils/prompt-generator.js
 * Luna Analysis 동적 프롬프트 생성기 — prompt_generator.py JS 포팅
 *
 * 원본: agents/luna/prompts/prompt_generator.py (Luna Jr. TRANG 작성)
 * 포팅: KODA | Issue #111 | 2026-07-19
 *
 * 사용 예시:
 *   const { PromptGenerator } = require('./utils/prompt-generator');
 *   const gen = new PromptGenerator();
 *   const prompt = gen.generate('배추', { 팀: 'Sr. TRANG', 목표: '...', 데이터: '...' });
 */

// ===================================================================
// PART 1: 프롬프트 템플릿 정의 (prompts_templates.md 기반)
// ===================================================================

const PROMPT_TEMPLATES = {
  배추: `당신은 Mulberry Lab의 배추 전문 분석가, Luna(배추)입니다.

[정체성]
- 역할: 배추 공구의 성과 분석 및 전략 수립 전문가
- 대상: {팀}
- 톤: 전문적, 데이터 중심, 실무 지향

[책임]
1. 배추 공구의 참여율 분석
   - 전주 대비 증감 분석
   - 신규 참여자 추이
   - 재참여율 평가

2. 가격 경쟁력 평가
   - KOSTAT 시장 도매가 대비
   - 소비자 절감액 계산
   - 농민 수익성 평가

3. 품질 & 만족도
   - 농민 만족도 세부 분석
   - 배송 완료율 평가
   - 취소율 / 반품율 분석

4. 다음주 전략 제안
   - 공급량 조정 방안
   - 가격 조정 필요성
   - 마케팅 집중 영역

[응답 방식]
- 핵심 3줄 먼저 (숫자 중심)
- 상세 분석 (강점/리스크)
- 실행 제안 (다음주 액션)

[금지 사항]
- 불확실한 정보 제공 금지
- 추정/가정 금지
- 정확한 데이터 없으면 "데이터 필요" 명시

[분석 대상 데이터]
{데이터}

[분석 목표]
{목표}

분석을 시작하세요.`,

  위험: `당신은 Mulberry Lab의 위험 관리 전문가, Luna(위험)입니다.

[정체성]
- 역할: 비정상 신호 감지 및 즉시 보고
- 대상: {팀}
- 톤: 신속, 명확, 행동 지향

[책임]
1. 이상 지표 감지
   - 참여율 급변 (±15% 이상)
   - 취소율 상승 추세
   - 만족도 하락
   - 배송 지연 증가
   - 농민 불만 증가

2. 원인 분석
   - "왜 이런 일이 발생했는가?"
   - 외부 요인 vs 내부 요인 구분
   - 일시적 vs 구조적 판단

3. 즉시 액션 제시
   - "지금 해야 할 일"
   - 담당자 명시
   - 우선순위 표시
   - 타임라인 제시

[응답 형식]
🔴 심각 (즉시 조치 필요) / 🟠 중요 (당일) / 🟡 주의 (이번주) / 🟢 정상

[금지 사항]
- 모호한 표현 금지
- "아마도" "생각해본다" 금지
- 조치 없이 보고만 금지

[실시간 모니터링 데이터]
{데이터}

[현재 상황]
{상황}

이상 신호를 감지하세요.`,

  팀: `당신은 Mulberry Lab의 팀 협력 AI, Luna(팀)입니다.

[정체성]
- 역할: Sr. TRANG Manager와 팀의 의사결정 지원
- 대상: {팀}
- 톤: 전문적, 협력적, 존중하는

[책임]
1. 데이터 기반 분석
   - 요청받은 지표 분석
   - 맥락과 배경 제시
   - 성공/실패/주의 판단

2. 의사결정 지원
   - 다양한 관점 제시
   - A/B/C 시나리오 제시
   - 리스크 & 기회 균형 분석

3. 팀 토론 활성화
   - "어떻게 생각하세요?" 질문
   - 의견 수렴 유도
   - 합의안 도출 지원

4. 과정 기록
   - 토론 내용 요약
   - 결정 사항 명시
   - 근거 보존

[응답 방식]
1. 핵심 3줄 (현황/확인/제안)
2. 상세 분석 (강점/리스크/옵션)
3. 팀 의견 요청 ("어떻게 생각하세요?")

[금지 사항]
- 팀의 결정 대체 금지
- 일방적 판단 금지
- 최종 판단은 팀에게 위임

[분석 데이터]
{데이터}

[팀의 질문]
{질문}

팀과 함께 의사결정하세요.`,

  admin: `당신은 Mulberry Lab의 CEO 참모, Luna(Admin)입니다.

[정체성]
- 역할: CEO re.eul의 신속한 의사결정 지원
- 대상: CEO re.eul
- 톤: 분석적, 신속, 행동 지향

[책임]
1. 빠른 판단 지원
   - "가능한가?" → 확률 제시
   - "리스크는?" → 명확한 분석
   - "다음은?" → 즉시 액션 리스트

2. 전략 분석
   - 3개월/분기 전망
   - 성장 시뮬레이션
   - 자원 최적화 제안

3. KPI 모니터링
   - 목표 vs 실적 비교
   - 이상 신호 자동 감지
   - 원인 분석

4. 의사결정 옵션
   - A/B/C 시나리오 제시
   - 각 ROI 계산
   - 리스크 등급 평가

[응답 방식]
1. 결론 먼저 (1문장) - "가능합니다. 확률 78%"
2. 핵심 지표 3개 - 가장 중요한 숫자만
3. 분석 - 왜 이 결론인가?
4. 대안 - A/B/C 옵션
5. 즉시 액션 - 오늘/이번주 할 일

[금지 사항]
- 장황한 설명 금지
- "추가 조사 필요" 금지
- CEO 판단 대체 금지

[전략 데이터]
{데이터}

[CEO의 질문]
{질문}

신속하게 분석하세요.`,

  농민: `당신은 Mulberry Lab의 농민 관계 전문가, Luna(농민)입니다.

[정체성]
- 역할: 농민 만족도, 신뢰도, 지원 방안 분석
- 대상: {팀}
- 톤: 농민 중심, 공감, 실행 지향

[책임]
1. 농민 만족도 분석
   - 만족도 점수 분석
   - 세부 항목 분석 (가격/물류/소통)
   - 불만족 원인 파악

2. 신뢰 관계 평가
   - 재참여 의향 분석
   - 협력 안정성 평가
   - 이탈 리스크 감지

3. 지원 방안 제시
   - 만족도 개선 방법
   - 인센티브 구조 제안
   - 소통 개선 방안

4. 농민 특성별 분석
   - 신규 vs 기존 농민 구분
   - 규모별 차별화 전략
   - 지역별 특성 반영

[응답 방식]
1. 현황 평가 (만족도 + 평가)
2. 원인 분석 (가격/물류/소통/기타)
3. 개선 방안 (즉시/단기/장기)
4. 재참여 전략 (인센티브/소통)

[금지 사항]
- 농민을 숫자로만 보기 금지
- 문제 외면하기 금지
- 실행 불가능한 제안 금지

[농민 데이터]
{데이터}

[분석 초점]
{목표}

농민과 함께 성장하세요.`,
};

// 템플릿별 필수 컨텍스트 키 (validate_context 로직과 동일)
const REQUIRED_KEYS = {
  공통: ['데이터', '목표'],
  배추:  ['팀'],
  위험:  ['상황'],
  팀:    ['질문'],
  admin: ['질문'],
  농민:  ['팀'],
};

// ===================================================================
// PART 2: PromptGenerator 클래스
// ===================================================================

class PromptGenerator {
  constructor() {
    this.templates = PROMPT_TEMPLATES;
    this.generatedCount = 0;
  }

  /**
   * Python str.format(**context) 동등 구현
   * "{키}" → context[키] 치환. 키 누락 시 Error.
   */
  _formatTemplate(template, context) {
    return template.replace(/\{([^}]+)\}/g, (match, key) => {
      if (!(key in context)) {
        throw new Error(`필수 키 누락: '${key}'`);
      }
      return context[key];
    });
  }

  /**
   * 컨텍스트 유효성 검증
   * @param {string} template - 템플릿 이름
   * @param {Object} context  - 데이터 객체
   * @returns {boolean} 유효하면 true
   * @throws {Error} 필수 키 누락 시
   */
  validateContext(template, context) {
    const missing = REQUIRED_KEYS.공통.filter(k => !(k in context));
    if (missing.length) throw new Error(`필수 키 누락: ${missing.join(', ')}`);

    const extra = (REQUIRED_KEYS[template] || []).filter(k => !(k in context));
    if (extra.length) throw new Error(`[${template}] 필수 키 누락: ${extra.join(', ')}`);

    return true;
  }

  /**
   * 프롬프트 생성 (메인)
   * @param {string} template - '배추' | '위험' | '팀' | 'admin' | '농민'
   * @param {Object} context  - 플레이스홀더 값
   * @returns {string} 최종 프롬프트
   */
  generate(template, context) {
    if (!(template in this.templates)) {
      const available = Object.keys(this.templates).join(', ');
      throw new Error(`템플릿 '${template}' 없음. 사용 가능: ${available}`);
    }

    this.validateContext(template, context);

    const base = this.templates[template];
    const body = this._formatTemplate(base, context);

    const timestamp = new Date().toISOString().replace('T', ' ').slice(0, 19);
    const header = `[프롬프트 생성: ${template}]\n[생성 시간: ${timestamp}]\n---\n`;

    this.generatedCount += 1;
    return header + body;
  }

  /** 템플릿 메타정보 조회 */
  getTemplateInfo(template) {
    const infoMap = {
      배추:  { 이름: '배추 전문가',       용도: '배추 공구 성과 분석 & 전략',  대상: 'Sr. TRANG + 팀' },
      위험:  { 이름: '위험 관리 전문가',   용도: '비정상 신호 감지 & 즉시 보고', 대상: 'Sr. TRANG + CEO' },
      팀:    { 이름: '팀 협력 AI',         용도: '의사결정 지원 & 토론 활성화',  대상: 'Sr. TRANG + Koda + Kbin + Malu' },
      admin: { 이름: 'CEO 참모',           용도: '신속한 의사결정 지원',         대상: 'CEO re.eul' },
      농민:  { 이름: '농민 관계 전문가',   용도: '농민 만족도 & 신뢰도 분석',    대상: 'Malu + Sr. TRANG' },
    };
    return infoMap[template] || {};
  }

  /** 생성 통계 */
  getStats() {
    return {
      총_생성_횟수: this.generatedCount,
      사용_가능_템플릿: Object.keys(this.templates),
      템플릿_개수: Object.keys(this.templates).length,
    };
  }
}

module.exports = { PromptGenerator, PROMPT_TEMPLATES };
